pkg/services/annotations/annotationsimpl/loki/alert_state_history_store.go

```diff
@@ -0,0 +1,399 @@
+package loki
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/grafana/grafana/pkg/components/simplejson"
+	"github.com/grafana/grafana/pkg/services/annotations/accesscontrol"
+
+	"github.com/grafana/grafana/pkg/infra/db"
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/services/ngalert/eval"
+	"github.com/grafana/grafana/pkg/services/ngalert/metrics"
+	ngmodels "github.com/grafana/grafana/pkg/services/ngalert/models"
+	"github.com/grafana/grafana/pkg/services/ngalert/state"
+	"github.com/grafana/grafana/pkg/services/ngalert/state/historian"
+	historymodel "github.com/grafana/grafana/pkg/services/ngalert/state/historian/model"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/grafana/grafana/pkg/services/annotations"
+	"github.com/grafana/grafana/pkg/setting"
+	"github.com/grafana/grafana/pkg/util/errutil"
+)
+
+// from grafana/pkg/services/ngalert/state/historian/loki.go
+const defaultQueryRange = 6 * time.Hour
+
+var (
+	ErrLokiStoreInternal    = errutil.NewBase(errutil.StatusInternal, "annotations.loki.internal")
+	ErrLokiStoreUnsupported = errutil.NotImplemented("annotations.loki.unsupported")
+	ErrMissingRule          = errors.New("missing rule")
+)
+
+type lokiQueryClient interface {
+	RangeQuery(ctx context.Context, query string, from, to, limit int64) (historian.QueryRes, error)
+}
+
+type AlertStateHistoryStore struct {
+	client lokiQueryClient
+	db     db.DB
+	log    log.Logger
+}
+
+func NewLokiAlertStateHistoryStore(cfg *setting.Cfg, db db.DB, log log.Logger) (*AlertStateHistoryStore, error) {
+	metrics := metrics.NewNGAlert(prometheus.DefaultRegisterer)
+
+	clientCfg, err := historian.NewLokiConfig(cfg.UnifiedAlerting.StateHistory)
+	if err != nil {
+		return nil, fmt.Errorf("invalid remote loki configuration: %w", err)
+	}
+	client := historian.NewLokiClient(clientCfg, historian.NewRequester(), metrics.GetHistorianMetrics(), log)
+
+	return &AlertStateHistoryStore{
+		client: client,
+		db:     db,
+		log:    log,
+	}, nil
+}
+
+func (r *AlertStateHistoryStore) Get(ctx context.Context, query *annotations.ItemQuery, accessResources accesscontrol.AccessResources) ([]*annotations.ItemDTO, error) {
+	items := make([]*annotations.ItemDTO, 0)
+	err := func() error {
+		var rule *ngmodels.AlertRule
+		var err error
+
+		if query.AlertID != 0 {
+			ruleQ := ruleQuery{ID: query.AlertID, OrgID: query.OrgID}
+			rule, err = r.getRule(ctx, ruleQ)
+			if err != nil {
+				return ErrLokiStoreInternal.Errorf("failed to fetch rule: %w", err)
+			}
+		}
+
+		histQuery := historyQueryFromItemQuery(ctx, query, rule, accessResources)
+
+		logQL, err := historian.BuildLogQuery(histQuery)
+		if err != nil {
+			return ErrLokiStoreInternal.Errorf("failed to build log query: %w", err)
+		}
+
+		now := time.Now().UTC()
+		if query.To == 0 {
+			query.To = now.UnixMilli()
```

Comment:
`from` and `to` are in UnixNano according to Loki's API, not UnixMilli. I wrote this bug once before - with Milli, the magnitude is much lower so Loki thinks the query is way in the past. Loki rejects the query due to it being older than the eviction time.

I would expect none of the queries hitting this code path to be accepted by Loki - how can we test this?

Reply:
Thanks for the callout on this 😁 I think the best way to test this is going to be integration tests against a real Loki instance, but I'm open to suggestions if you feel like there's a faster way to do it.

Reply:
We discussed this more and there's not a good way for reviewers to test as-is. I'm going to merge the link between this and the annotation service first so reviewers can actually test this PR themselves since it's the bulk of the logic change.